### PR TITLE
Add EVM Tests 

### PR DESCRIPTION
**Pull request by ryardley**
Closes: #169 

This provides the ability to test evm event subscription. 

- You need to have `anvil` installed in the PATH in order to run the tests
- This is important in order to enable historical events to be consumed and tested.
- Update `alloy` to a modern version (`5.2`)
- Added a `EnclaveEvent::TestEvent` we can use in testing to test event ordering and transmission issues.
- Added fixture `emit_logs.sol` as well as a build script to build the artifact used in the rust test
- It adds generic Transport support to `WithChainId` which was required to support providers that didn't use `BoxTransport` such as the default anvil transports provided by alloy (` ProviderBuilder::new().with_recommended_fillers().on_anvil_with_wallet()`)
- Ensure `solc` is available as a dependency 
- New development pre-requisites [documented here](https://github.com/gnosisguild/enclave/blob/ry/169-add-tests-consolidated/packages/ciphernode/README.md)

---
- [x] Can use ` ProviderBuilder::new().with_recommended_fillers().on_anvil_with_wallet()`
- [x] Basic EVM event test works 
- [x] Ensure Anvil available in CI to enable testing on CI not to fail
- [x] Documentation
- [x] Tidy up rabbit review suggestions

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `TestEvent` type for enhanced event handling.
  - Added `build_fixtures.sh` script to automate Ethereum smart contract fixture building.
  - New `EmitLogs` Solidity contract for event logging capabilities.

- **Improvements**
  - Updated CI workflow to install Foundry and Solidity compiler.
  - Modified test execution command for better integration with Yarn.
  - Enhanced type safety and flexibility in event handling and provider management.
  - Added a "Prerequisites" section in the README for required tool versions.

- **Bug Fixes**
  - Adjusted method signatures to ensure compatibility with updated provider types.

- **Tests**
  - Added tests for logging events using the Actix framework and Alloy library. 

- **Chores**
  - Updated `.gitignore` to exclude JSON files in test fixtures.
  - Introduced `test.sh` script for streamlined testing processes.

### CODE REVIEW

File: packages/ciphernode/evm/tests/fixtures/emit_logs.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
+pragma solidity >=0.4.24;
+
+contract EmitLogs {
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/tests/fixtures/emit_logs.sol:3]: Not important here

File: packages/ciphernode/evm/tests/fixtures/emit_logs.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    event ValueChanged(address indexed author, string oldValue, string newValue);
+
+    string _value;
+
+    constructor() {
+        _value = "";
+    }
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/tests/fixtures/emit_logs.sol:11]: not important

File: packages/ciphernode/evm/tests/evm_reader.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+use enclave_core::{EnclaveEvent, EventBus, GetHistory, TestEvent};
+use evm::{helpers::WithChainId, EvmEventReader};
+use std::{sync::Arc, time::Duration};
+use tokio::time::sleep;
+
+sol!(
+    #[sol(rpc)]
+    EmitLogs,
+    "tests/fixtures/emit_logs.json"
+);
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/tests/evm_reader.rs:18]: This is built

**Review by auryn-macmillan — APPROVED** [review]
lgtm

File: packages/ciphernode/evm/tests/evm_reader.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        .await?;
+
+    contract
+        .setValue("world!".to_string())
+        .send()
+        .await?
+        .watch()
+        .await?;
+
+    sleep(Duration::from_millis(1)).await;
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/tests/evm_reader.rs:63]: No it is not

File: packages/ciphernode/evm/tests/evm_reader.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        .filter_map(|evt| match evt {
+            EnclaveEvent::TestEvent { data, .. } => Some(data.msg),
+            _ => None,
+        })
+        .collect();
+
+    assert_eq!(msgs, vec!["hello", "world!"]);
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (ryardley) [reply packages/ciphernode/evm/tests/evm_reader.rs:80]: Anvil is killed when the value is dropped

**Pull request merged**